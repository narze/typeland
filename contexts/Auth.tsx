import React, { useEffect, useReducer } from 'react'
import { auth as firebaseAuth } from '../config/firebase'
import firebase from 'firebase'

interface AuthState {
  user: firebase.User | null
}

interface AuthContextProps {
  state: AuthState
  dispatch: React.Dispatch<any>
}

interface AuthProviderProps {
  children: React.ReactNode
  value?: AuthContextProps
}

const defaultAuth: AuthState = {
  user: null,
}

type Action = { type: 'setUser'; user: firebase.User } | { type: 'unsetUser' }

export const reducer = (state: AuthState, action: Action): AuthState => {
  switch (action.type) {
    case 'setUser':
      return {
        ...state,
        user: action.user,
      }
    case 'unsetUser':
      return {
        ...state,
        user: null,
      }
    default:
      return state
  }
}

export const AuthContext = React.createContext({} as AuthContextProps)

export const AuthProvider: React.FC<AuthProviderProps> = ({
  children,
  value,
}) => {
  const [state, dispatch] = useReducer(reducer, value.state || defaultAuth)

  useEffect(() => {
    const unsubscribe = firebaseAuth.onAuthStateChanged((authUser) => {
      if (authUser) {
        //The user is logged in
        dispatch({
          type: 'setUser',
          user: authUser,
        })
      } else {
        //The user is logged out
        dispatch({
          type: 'unsetUser',
        })
      }
    })
    return () => {
      // Any clean up operation goes in here
      unsubscribe()
    }
  }, [])

  return (
    <AuthContext.Provider value={{ state, dispatch }}>
      {children}
    </AuthContext.Provider>
  )
}
