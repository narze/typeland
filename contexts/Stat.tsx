import React from 'react'

const stats = {
  correct: 0,
  wrong: 0,
  total: 0,
}

export const StatContext = React.createContext(stats)
