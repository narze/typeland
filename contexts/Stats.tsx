import React, { useContext, useReducer } from 'react'

interface State {
  correct: number
  wrong: number
  total: number
}

type Dispatch = React.Dispatch<Action>

interface StatsProviderProps {
  children: React.ReactNode
  initialState?: State
}

export type Action =
  | { type: 'incrementCorrect' }
  | { type: 'incrementWrong' }
  | { type: 'reset' }

const defaultState: State = {
  correct: 0,
  wrong: 0,
  total: 0,
}

function reducer(state: State, action: Action): State {
  switch (action.type) {
    case 'incrementCorrect':
      return {
        ...state,
        correct: state.correct + 1,
        total: state.total + 1,
      }
    case 'incrementWrong':
      return {
        ...state,
        wrong: state.wrong + 1,
        total: state.total + 1,
      }
    case 'reset':
      return { ...defaultState }
    default:
      throw new Error(`Unhandled action type}`)
  }
}

const StatsStateContext = React.createContext<State | undefined>(undefined)
const StatsDispatchContext = React.createContext<Dispatch | undefined>(
  undefined
)

StatsStateContext.displayName = 'Stats'

const StatsStateConsumer = StatsStateContext.Consumer

function useStatsState(): State {
  const context = useContext(StatsStateContext)
  if (context === undefined) {
    throw new Error('useStatsState must be used within a StatsProvider')
  }
  return context
}

function useStatsDispatch(): Dispatch {
  const context = useContext(StatsDispatchContext)
  if (context === undefined) {
    throw new Error('useStatsDispatch must be used within a StatsProvider')
  }
  return context
}

function useStats(): [State, Dispatch] {
  return [useStatsState(), useStatsDispatch()]
}

const StatsProvider: React.FC<StatsProviderProps> = ({
  children,
  initialState = defaultState,
}) => {
  const [state, dispatch] = useReducer(reducer, initialState)

  return (
    <StatsStateContext.Provider value={state}>
      <StatsDispatchContext.Provider value={dispatch}>
        {children}
      </StatsDispatchContext.Provider>
    </StatsStateContext.Provider>
  )
}

export {
  reducer,
  StatsProvider,
  StatsStateConsumer,
  useStats,
  useStatsDispatch,
  useStatsState,
}
