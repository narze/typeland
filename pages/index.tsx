/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import tw from '@tailwindcssinjs/macro'
import { useEffect, useState } from 'react'
import { TypingArea } from '../components/TypingArea'
import { randomWords } from '../utils/wordsDb'
import { useStats } from '../contexts/Stats'
import { useAuth } from '@/contexts/Auth'
import Link from 'next/link'
import { auth as firebaseAuth } from '../config/firebase'

const s = {
  container: tw`
    h-screen min-w-full
    flex
    items-center justify-center
  `,
  main: [
    tw`
      max-w-1/2
    `,
    css`
      min-width: 640px;
    `,
  ],
  title: tw`
    text-4xl
    text-blue-600
    antialiased
  `,
  typingInput: tw`
    h-0
    w-0
  `,
  footer: tw`
    fixed
    bottom-0
    pb-8
    text-center
  `,
  header: tw`
    fixed
    top-0
    right-8
    pt-4
    text-right
  `,
  githubLink: tw`
    text-xs
  `,
  result: tw`
    mt-4
    text-center
    text-xl
  `,
  restart: tw`
    mt-2
    text-center
  `,
  restartButton: tw`
    py-2 px-4
    bg-transparent hover:bg-blue-500
    font-semibold
    hover:text-white
    border border-blue-500 hover:border-transparent
    rounded
  `,
  restartHint: tw`
    mt-1
    text-xs
  `,
  liveWpm: tw`
    text-base
  `,
}

export const Home = (): JSX.Element => {
  const [words, setWords] = useState([''])
  const [userTypeInput, setUserTypeInput] = useState([''])
  const [inputIsFocused, setInputIsFocused] = useState(true)
  const [finished, setFinished] = useState(false)
  const [started, setStarted] = useState(false)
  const [startTime, setStartTime] = useState(0)
  const [finishTime, setFinishTime] = useState(0)
  const [wpm, setWpm] = useState(0)
  const [liveWpm, setLiveWpm] = useState(0)
  const [elapsedMs, setElapsedMs] = useState(0)
  const [promptRestart, setPromptRestart] = useState(false)
  const [stats, dispatch] = useStats()
  const [auth] = useAuth()

  const DEFAULT_WORD_COUNT = 30
  const TIMER_LOOP_MS = 1000

  useEffect(() => {
    if (window._seed) {
      setWords(window._seed.words.split(' '))
    } else {
      setWords(randomWords(DEFAULT_WORD_COUNT))
    }
  }, [])

  useEffect(() => {
    let timer

    if (started) {
      timer = setInterval(() => {
        setElapsedMs((ms) => ms + TIMER_LOOP_MS)
      }, TIMER_LOOP_MS)
    } else {
      setElapsedMs(0)
      clearInterval(timer)
    }

    return () => clearInterval(timer)
  }, [started])

  useEffect(() => {
    setLiveWpm(Math.round((stats.correct * 60) / (elapsedMs / 1000.0)) || 0)
  }, [elapsedMs])

  const handleDelete = ({ word = false }) => {
    const newUserTypeInput = [...userTypeInput]

    if (word) {
      newUserTypeInput[newUserTypeInput.length - 1] = ''
    } else {
      newUserTypeInput[newUserTypeInput.length - 1] = newUserTypeInput[
        newUserTypeInput.length - 1
      ].slice(0, -1)
    }

    setUserTypeInput(newUserTypeInput)
  }

  const handleSpace = () => {
    if (!userTypeInput[userTypeInput.length - 1].length) {
      return
    }
    if (userTypeInput.length == words.length) {
      setFinished(true)
      return
    }

    const newUserTypeInput = [...userTypeInput]

    newUserTypeInput[newUserTypeInput.length] = ''

    setUserTypeInput(newUserTypeInput)
  }

  const handleEnter = () => {
    if (finished || promptRestart) {
      return restart()
    }

    if (started) {
      setPromptRestart(true)
    }
  }

  const handleType = (e) => {
    const { key, altKey, ctrlKey, metaKey } = e

    if (key == 'Backspace') {
      return handleDelete({ word: altKey || ctrlKey || metaKey })
    }

    if (key == 'Enter') {
      return handleEnter()
    }

    // Filter out modifiers
    if (key.length != 1 || altKey || ctrlKey || metaKey) {
      return
    }

    if (promptRestart) {
      setPromptRestart(false)
    }

    if (key == ' ') {
      return handleSpace()
    }

    if (!started) {
      setStarted(true)
    }

    const newUserTypeInput = [...userTypeInput]
    newUserTypeInput[newUserTypeInput.length - 1] = newUserTypeInput[
      newUserTypeInput.length - 1
    ].concat(key)
    setUserTypeInput(newUserTypeInput)

    if (userTypeInput.length == words.length) {
      if (
        newUserTypeInput[newUserTypeInput.length - 1] == words[words.length - 1]
      ) {
        setFinished(true)
      }
    }
  }

  const restart = () => {
    setWords(randomWords(DEFAULT_WORD_COUNT))
    setUserTypeInput([''])
    setInputIsFocused(true)
    setFinished(false)
    setStarted(false)
    setStartTime(0)
    setFinishTime(0)
    setWpm(0)
    setLiveWpm(0)
    setElapsedMs(0)
    setPromptRestart(false)
    dispatch({ type: 'reset' })
  }

  useEffect(() => {
    if (started && !finished) {
      setStartTime(+new Date())
    }

    if (started && finished) {
      setFinishTime(+new Date())
    }
  }, [started, finished])

  useEffect(() => {
    if (startTime && finishTime && !wpm) {
      setWpm(
        Math.round((stats.correct * 60) / ((finishTime - startTime) / 1000.0))
      )
    }
  }, [startTime, finishTime, wpm])

  return (
    <div
      css={s.container}
      onClick={() => document.getElementById('typingInput').focus()}
    >
      <div css={s.header}>
        {auth.user ? (
          <>
            <span>{auth.user.email}</span>
            <button
              css={tw`ml-4 cursor-pointer`}
              onClick={() => firebaseAuth.signOut()}
            >
              Logout
            </button>
          </>
        ) : (
          <Link href="/login">Login</Link>
        )}
      </div>

      <main css={s.main}>
        <h1 css={s.title}>Typeland</h1>

        <TypingArea
          words={words}
          userWords={userTypeInput}
          showCaret={inputIsFocused}
          finished={finished}
        />

        {started && !finished && <div css={s.liveWpm}>{liveWpm} wpm</div>}

        {started && (
          <div css={s.liveWpm}>
            Stats (correct/wrong/total) : {stats.correct}/{stats.wrong}/
            {stats.total}
          </div>
        )}

        {finished && finishTime && (
          <>
            <div css={s.result}>Good job! {wpm} wpm</div>
            <div css={s.restart}>
              <button css={s.restartButton} onClick={restart}>
                Restart
              </button>
              <div css={s.restartHint}>
                Click or press <code>Enter</code>
              </div>
            </div>
          </>
        )}

        {promptRestart && (
          <>
            <div css={s.restart}>
              <button css={s.restartButton} onClick={restart}>
                Restart
              </button>
              <div css={s.restartHint}>
                Press <code>Enter</code> again to restart typing.
              </div>
            </div>
          </>
        )}

        <input
          id="typingInput"
          data-testid="typingInput"
          css={s.typingInput}
          type="text"
          ref={(i) => i && i.focus()} // Autofocus
          onKeyDown={handleType}
          onBlur={() => {
            setInputIsFocused(false)
          }}
          onFocus={() => {
            setInputIsFocused(true)
          }}
          tabIndex={0}
        />
      </main>

      <div css={s.footer}>
        <a
          css={s.githubLink}
          href="https://github.com/narze/typeland"
          target="_blank"
          rel="noreferrer"
        >
          GitHub
        </a>
      </div>
    </div>
  )
}

export default Home
