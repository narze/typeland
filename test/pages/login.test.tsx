import React from 'react'
import { render } from '../testUtils'
import { Login } from '@/pages/login'
import { AuthProvider } from '@/contexts/Auth'

const mockRouterPush = jest.fn()

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/',
      pathname: '',
      query: '',
      asPath: '',
      push: mockRouterPush,
    }
  },
}))

const renderWithAuthProvider = (
  ui,
  { providerProps = {}, ...renderOptions } = {}
) => {
  return render(
    <AuthProvider {...providerProps}>{ui}</AuthProvider>,
    renderOptions
  )
}

describe('Login page', () => {
  it('renders sign up form', () => {
    const { asFragment, getByText } = renderWithAuthProvider(<Login />, {})
    getByText(/Sign Up/i)

    expect(asFragment()).toMatchSnapshot()
  })

  describe('when clicking Sign up', () => {
    it('renders user email', () => {
      const { asFragment, getByRole } = renderWithAuthProvider(<Login />, {})
      getByRole('button', { name: /Sign Up/i })

      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('when user is set in state (already logged in)', () => {
    it('redirects to home page', () => {
      const user = { email: 'foo@bar.com' }
      const providerProps = { initialState: { user } }

      const { asFragment } = renderWithAuthProvider(<Login />, {
        providerProps,
      })

      expect(mockRouterPush).toBeCalledTimes(1)
      expect(mockRouterPush).toBeCalledWith('/')

      expect(asFragment()).toMatchSnapshot()

      mockRouterPush.mockClear()
    })
  })
})
