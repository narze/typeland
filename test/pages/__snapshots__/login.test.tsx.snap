// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page renders sign up form 1`] = `
<DocumentFragment>
  .emotion-13 {
  height: 100vh;
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-1 {
  position: fixed;
  top: 0;
  right: 2rem;
  padding-top: 1rem;
  text-align: right;
}

.emotion-0 {
  margin-left: 1rem;
  cursor: pointer;
}

.emotion-12 {
  max-width: 50%;
  min-width: 640px;
}

.emotion-2 {
  --text-opacity: 1;
  font-size: 2.25rem;
  color: #1c64f2;
  color: rgba(28,100,242,var(--text-opacity));
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-3 {
  font-size: 1rem;
  padding-right: 12px;
  padding-bottom: 4px;
  opacity: 1;
  font-weight: 500;
  text-align: left;
  vertical-align: middle;
  display: inline-block;
}

.emotion-4 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  height: 2.5rem;
  border-radius: 0.25rem;
  border: 1px solid;
  border-color: inherit;
  background-color: #fff;
}

.emotion-4:hover {
  border-color: #CBD5E0;
}

.emotion-4:focus {
  z-index: 1;
  border-color: #3182ce;
  box-shadow: 0 0 0 1px #3182ce;
}

.emotion-4:disabled,
.emotion-4:disabled:focus,
.emotion-4:disabled:hover,
.emotion-4[aria-disabled=true],
.emotion-4[aria-disabled=true]:focus,
.emotion-4[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-4[aria-invalid=true] {
  border-color: #e53e3e;
  box-shadow: 0 0 0 1px #e53e3e;
}

.emotion-4[aria-readonly=true],
.emotion-4[readonly] {
  background-color: transparent;
  box-shadow: none !important;
  -webkit-user-select: all;
  -moz-user-select: all;
  -ms-user-select: all;
  user-select: all;
}

.emotion-11 {
  margin-top: 1rem;
}

.emotion-9 {
  border-radius: 0.25rem;
  font-weight: 600;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  line-height: 1.2;
  outline: none;
  height: 2.5rem;
  min-width: 2.5rem;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  background-color: #319795;
  color: #fff;
}

.emotion-9:hover {
  background-color: #2C7A7B;
}

.emotion-9:focus {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-9:active,
.emotion-9[data-active=true] {
  background-color: #285E61;
}

.emotion-9:disabled,
.emotion-9:disabled:focus,
.emotion-9:disabled:hover,
.emotion-9[aria-disabled=true],
.emotion-9[aria-disabled=true]:focus,
.emotion-9[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

.emotion-10 {
  border-radius: 0.25rem;
  font-weight: 600;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  line-height: normal;
  outline: none;
  height: auto;
  min-width: 2.5rem;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding: 0;
  color: #718096;
  margin-left: 1rem;
}

.emotion-10:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-10:focus {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-10:active,
.emotion-10[data-active=true] {
  color: #2D3748;
}

.emotion-10:disabled,
.emotion-10:disabled:focus,
.emotion-10:disabled:hover,
.emotion-10[aria-disabled=true],
.emotion-10[aria-disabled=true]:focus,
.emotion-10[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

<div
    class="emotion-13"
  >
    <div
      class="emotion-1"
    >
      <button
        class="emotion-0"
      >
        Back
      </button>
    </div>
    <main
      class="emotion-12"
    >
      <h1
        class="emotion-2"
      >
        Login
      </h1>
      <form>
        <div
          class="emotion-5"
          role="group"
        >
          <label
            class="emotion-3"
            for="email"
          >
            Email
          </label>
          <input
            class="emotion-4"
            id="email"
            name="email"
          />
        </div>
        <div
          class="emotion-5"
          role="group"
        >
          <label
            class="emotion-3"
            for="password"
          >
            Password
          </label>
          <input
            class="emotion-4"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <div
          class="emotion-11"
        >
          <button
            aria-disabled="false"
            class="emotion-9"
            type="submit"
          >
            Login
          </button>
          <button
            class="emotion-10"
            type="button"
          >
            Sign Up
          </button>
        </div>
      </form>
    </main>
  </div>
</DocumentFragment>
`;

exports[`Login page when clicking Sign up renders user email 1`] = `
<DocumentFragment>
  .emotion-13 {
  height: 100vh;
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-1 {
  position: fixed;
  top: 0;
  right: 2rem;
  padding-top: 1rem;
  text-align: right;
}

.emotion-0 {
  margin-left: 1rem;
  cursor: pointer;
}

.emotion-12 {
  max-width: 50%;
  min-width: 640px;
}

.emotion-2 {
  --text-opacity: 1;
  font-size: 2.25rem;
  color: #1c64f2;
  color: rgba(28,100,242,var(--text-opacity));
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-3 {
  font-size: 1rem;
  padding-right: 12px;
  padding-bottom: 4px;
  opacity: 1;
  font-weight: 500;
  text-align: left;
  vertical-align: middle;
  display: inline-block;
}

.emotion-4 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  height: 2.5rem;
  border-radius: 0.25rem;
  border: 1px solid;
  border-color: inherit;
  background-color: #fff;
}

.emotion-4:hover {
  border-color: #CBD5E0;
}

.emotion-4:focus {
  z-index: 1;
  border-color: #3182ce;
  box-shadow: 0 0 0 1px #3182ce;
}

.emotion-4:disabled,
.emotion-4:disabled:focus,
.emotion-4:disabled:hover,
.emotion-4[aria-disabled=true],
.emotion-4[aria-disabled=true]:focus,
.emotion-4[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-4[aria-invalid=true] {
  border-color: #e53e3e;
  box-shadow: 0 0 0 1px #e53e3e;
}

.emotion-4[aria-readonly=true],
.emotion-4[readonly] {
  background-color: transparent;
  box-shadow: none !important;
  -webkit-user-select: all;
  -moz-user-select: all;
  -ms-user-select: all;
  user-select: all;
}

.emotion-11 {
  margin-top: 1rem;
}

.emotion-9 {
  border-radius: 0.25rem;
  font-weight: 600;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  line-height: 1.2;
  outline: none;
  height: 2.5rem;
  min-width: 2.5rem;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  background-color: #319795;
  color: #fff;
}

.emotion-9:hover {
  background-color: #2C7A7B;
}

.emotion-9:focus {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-9:active,
.emotion-9[data-active=true] {
  background-color: #285E61;
}

.emotion-9:disabled,
.emotion-9:disabled:focus,
.emotion-9:disabled:hover,
.emotion-9[aria-disabled=true],
.emotion-9[aria-disabled=true]:focus,
.emotion-9[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

.emotion-10 {
  border-radius: 0.25rem;
  font-weight: 600;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  line-height: normal;
  outline: none;
  height: auto;
  min-width: 2.5rem;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding: 0;
  color: #718096;
  margin-left: 1rem;
}

.emotion-10:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-10:focus {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-10:active,
.emotion-10[data-active=true] {
  color: #2D3748;
}

.emotion-10:disabled,
.emotion-10:disabled:focus,
.emotion-10:disabled:hover,
.emotion-10[aria-disabled=true],
.emotion-10[aria-disabled=true]:focus,
.emotion-10[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

<div
    class="emotion-13"
  >
    <div
      class="emotion-1"
    >
      <button
        class="emotion-0"
      >
        Back
      </button>
    </div>
    <main
      class="emotion-12"
    >
      <h1
        class="emotion-2"
      >
        Login
      </h1>
      <form>
        <div
          class="emotion-5"
          role="group"
        >
          <label
            class="emotion-3"
            for="email"
          >
            Email
          </label>
          <input
            class="emotion-4"
            id="email"
            name="email"
          />
        </div>
        <div
          class="emotion-5"
          role="group"
        >
          <label
            class="emotion-3"
            for="password"
          >
            Password
          </label>
          <input
            class="emotion-4"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <div
          class="emotion-11"
        >
          <button
            aria-disabled="false"
            class="emotion-9"
            type="submit"
          >
            Login
          </button>
          <button
            class="emotion-10"
            type="button"
          >
            Sign Up
          </button>
        </div>
      </form>
    </main>
  </div>
</DocumentFragment>
`;

exports[`Login page when user is set in state (already logged in) redirects to home page 1`] = `
<DocumentFragment>
  .emotion-13 {
  height: 100vh;
  min-width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-1 {
  position: fixed;
  top: 0;
  right: 2rem;
  padding-top: 1rem;
  text-align: right;
}

.emotion-0 {
  margin-left: 1rem;
  cursor: pointer;
}

.emotion-12 {
  max-width: 50%;
  min-width: 640px;
}

.emotion-2 {
  --text-opacity: 1;
  font-size: 2.25rem;
  color: #1c64f2;
  color: rgba(28,100,242,var(--text-opacity));
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

.emotion-3 {
  font-size: 1rem;
  padding-right: 12px;
  padding-bottom: 4px;
  opacity: 1;
  font-weight: 500;
  text-align: left;
  vertical-align: middle;
  display: inline-block;
}

.emotion-4 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  height: 2.5rem;
  border-radius: 0.25rem;
  border: 1px solid;
  border-color: inherit;
  background-color: #fff;
}

.emotion-4:hover {
  border-color: #CBD5E0;
}

.emotion-4:focus {
  z-index: 1;
  border-color: #3182ce;
  box-shadow: 0 0 0 1px #3182ce;
}

.emotion-4:disabled,
.emotion-4:disabled:focus,
.emotion-4:disabled:hover,
.emotion-4[aria-disabled=true],
.emotion-4[aria-disabled=true]:focus,
.emotion-4[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-4[aria-invalid=true] {
  border-color: #e53e3e;
  box-shadow: 0 0 0 1px #e53e3e;
}

.emotion-4[aria-readonly=true],
.emotion-4[readonly] {
  background-color: transparent;
  box-shadow: none !important;
  -webkit-user-select: all;
  -moz-user-select: all;
  -ms-user-select: all;
  user-select: all;
}

.emotion-11 {
  margin-top: 1rem;
}

.emotion-9 {
  border-radius: 0.25rem;
  font-weight: 600;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  line-height: 1.2;
  outline: none;
  height: 2.5rem;
  min-width: 2.5rem;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  background-color: #319795;
  color: #fff;
}

.emotion-9:hover {
  background-color: #2C7A7B;
}

.emotion-9:focus {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-9:active,
.emotion-9[data-active=true] {
  background-color: #285E61;
}

.emotion-9:disabled,
.emotion-9:disabled:focus,
.emotion-9:disabled:hover,
.emotion-9[aria-disabled=true],
.emotion-9[aria-disabled=true]:focus,
.emotion-9[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

.emotion-10 {
  border-radius: 0.25rem;
  font-weight: 600;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-transition: all 250ms;
  transition: all 250ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  white-space: nowrap;
  vertical-align: middle;
  line-height: normal;
  outline: none;
  height: auto;
  min-width: 2.5rem;
  font-size: 1rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding: 0;
  color: #718096;
  margin-left: 1rem;
}

.emotion-10:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-10:focus {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-10:active,
.emotion-10[data-active=true] {
  color: #2D3748;
}

.emotion-10:disabled,
.emotion-10:disabled:focus,
.emotion-10:disabled:hover,
.emotion-10[aria-disabled=true],
.emotion-10[aria-disabled=true]:focus,
.emotion-10[aria-disabled=true]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

<div
    class="emotion-13"
  >
    <div
      class="emotion-1"
    >
      <button
        class="emotion-0"
      >
        Back
      </button>
    </div>
    <main
      class="emotion-12"
    >
      <h1
        class="emotion-2"
      >
        Login
      </h1>
      <form>
        <div
          class="emotion-5"
          role="group"
        >
          <label
            class="emotion-3"
            for="email"
          >
            Email
          </label>
          <input
            class="emotion-4"
            id="email"
            name="email"
          />
        </div>
        <div
          class="emotion-5"
          role="group"
        >
          <label
            class="emotion-3"
            for="password"
          >
            Password
          </label>
          <input
            class="emotion-4"
            id="password"
            name="password"
            type="password"
          />
        </div>
        <div
          class="emotion-11"
        >
          <button
            aria-disabled="false"
            class="emotion-9"
            type="submit"
          >
            Login
          </button>
          <button
            class="emotion-10"
            type="button"
          >
            Sign Up
          </button>
        </div>
      </form>
    </main>
  </div>
</DocumentFragment>
`;
